### PR TITLE
[TT-7808] In proxy-only mode, GraphQL query return errors from upstream

### PR DESCRIPTION
**Pull request by buraksezer**
PR for [TT-7808](https://tyktech.atlassian.net/browse/TT-7808)

We pass the upstream headers and status codes to the client in proxy-only mode. This is an existing feature. With this PR, we start attaching the upstream's error messages to the response. 

For example, say that a GraphQL service returns the following, non-standard error message with HTTP 401 status code:

```json
{"message": "Unauthorized"}
```

We return the following error message with HTTP 401 message to the client:

```json
{
    "errors": [
        {
            "message": "unable to resolve",
            "locations": [
                {
                    "line": 2,
                    "column": 5
                }
            ]
        }
    ],
    "data": null
}
```

It was not possible to reach out to the original error message from upstream. With this PR, we started returning the following error message with HTTP 401 status code:

```json
{
    "errors": [
        {
            "message": "unable to resolve",
            "locations": [
                {
                    "line": 2,
                    "column": 5
                }
            ],
            "extensions": {
                "message": "Unauthorized"
            }
        }
    ],
    "data": null
}
```

A new configuration item has been added to the API definition(`graphql.proxy.return_errors_from_upstream`). This solution is only valid for proxy-only mode and we only keep the original response if upstream returns an error(status code >= 400).

[TT-7808]: https://tyktech.atlassian.net/browse/TT-7808?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5049/merge
Commit:  
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5057718977)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5049/merge
Commit: 537ad345a516bf11db257a5f83a6bd31758e822f 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5057774099)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5049/merge
Commit: b4b95066c0d6adf0f18a7cfddd6cad9be864539c 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5057884091)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5049/merge
Commit: 623946e32f71fac5323d74a7a8fba5b7b6c8ec62 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5066510082)

File: apidef/api_definitions.go [hunk 1]
```diff
@@ -749,9 +749,10 @@ const (
 )
 
 type GraphQLProxyConfig struct {
-	AuthHeaders      map[string]string `bson:"auth_headers" json:"auth_headers"`
-	SubscriptionType SubscriptionType  `bson:"subscription_type" json:"subscription_type,omitempty"`
-	RequestHeaders   map[string]string `bson:"request_headers" json:"request_headers"`
+	AuthHeaders              map[string]string `bson:"auth_headers" json:"auth_headers"`
+	SubscriptionType         SubscriptionType  `bson:"subscription_type" json:"subscription_type,omitempty"`
+	RequestHeaders           map[string]string `bson:"request_headers" json:"request_headers"`
+	ReturnErrorsFromUpstream bool              `bson:"return_errors_from_upstream" json:"return_errors_from_upstream"`
```
Comment (pvormste) [inline apidef/api_definitions.go:755]: To make this feature proof maybe we can create a new struct here:
```go
UseResponseExtensions GraphQLResponseExtensions `bson:"use_response_extensions" json:"use_response_extensions"`

type GraphQLResponseExtensions struct {
    OnErrorForwarding bool `bson:"on_error_forwarding" json:"on_error_forwarding"`
}
```

which results in
```json
"proxy": {
        "auth_headers": {},
        "request_headers": null,
        "use_response_extensions": {
             "on_error_forwarding": true
        }
    },
```
Feel free to adjust this idea

**Review by pvormste — CHANGES_REQUESTED** [review]

File: gateway/mw_graphql_transport.go [hunk 2]
```diff
@@ -81,6 +84,23 @@ func (g *GraphQLEngineTransport) handleProxyOnly(proxyOnlyCtx *GraphQLProxyOnlyC
...
+			return nil, err
+		}
+		defer func() {
+			_ = response.Body.Close()
+		}()
+		// graphql-go-tools uses response.body to resolve the upstream response.
+		// It's not possible to re-use io.ReadCloser. Because of that, we keep the
+		// original error message for later use.
+		// See TT-7808
+		proxyOnlyCtx.originalResponseBody = bytes.NewReader(body)
```
Comment (titpetric) [inline gateway/mw_graphql_transport.go:101]: Can you consider using `nopCloserBuffer` or `nopCloser`, both exist in reverse_proxy.go, and could be isolated and moved into /internal/httputil; they allow for re-reading of an io.ReadCloser, which should match request or response bodies.

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5049/merge
Commit: 9fe85398d78d2f5656c618a61983169c17d537b4 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5068820130)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5049/merge
Commit: 04bb5d34aef2bf0cc1242c125fd6cf14ef41b528 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5068818825)

**Review by pvormste — APPROVED** [review]

**Review by kofoworola — APPROVED** [review]

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/5049/merge
Commit: d8f896aadcd9dfc653edb0ccd7681f5d2920f80b 
Triggered by: pull_request (@buraksezer)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/5069773944)

**Pull request merged**